modules/byDataAnalyticsAdapter.js

```diff
@@ -0,0 +1,309 @@
+          regexv = new RegExp(data[i].version + '[- /:;]([\\d._]+)', 'i');
+          matches = string.match(regexv);
+          version = '';
+          if (matches) { if (matches[1]) { matches = matches[1]; } }
+          if (matches) {
+            matches = matches.split(/[._]+/);
+            for (j = 0; j < matches.length; j += 1) {
+              if (j === 0) {
+                version += matches[j] + '.';
+              } else {
+                version += matches[j];
+              }
+            }
+          } else {
+            version = '0';
+          }
+          return {
+            name: data[i].name,
+            version: parseFloat(version)
+          };
+        }
+      }
+      return { name: 'unknown', version: 0 };
+    }
+  };
+  function generateUid() {
+    try {
+      var buffer = new Uint8Array(16);
+      crypto.getRandomValues(buffer);
+      buffer[6] = (buffer[6] & ~176) | 64;
+      buffer[8] = (buffer[8] & ~64) | 128;
+      var hex = Array.prototype.map.call(new Uint8Array(buffer), function(x) {
+        return ('00' + x.toString(16)).slice(-2);
+      }).join('');
+      return hex.slice(0, 5) + '-' + hex.slice(5, 9) + '-' + hex.slice(9, 13) + '-' + hex.slice(13, 18);
+    } catch (e) {
+      return '';
+    }
+  }
+  function base64url(source) {
+    var encodedSource = Base64.stringify(source);
+    encodedSource = encodedSource.replace(/=+$/, '');
+    encodedSource = encodedSource.replace(/\+/g, '-');
+    encodedSource = encodedSource.replace(/\//g, '_');
+    return encodedSource;
+  }
+  function getJWToken(data) {
+    var header = {
+      'alg': 'HS256',
+      'typ': 'JWT'
+    };
+    var stringifiedHeader = enc.parse(JSON.stringify(header));
+    var encodedHeader = base64url(stringifiedHeader);
+    var stringifiedData = enc.parse(JSON.stringify(data));
+    var encodedData = base64url(stringifiedData);
+    var token = encodedHeader + '.' + encodedData;
+    var signature = hmacSHA512(token, secretKey);
+    signature = base64url(signature);
+    var signedToken = token + '.' + signature;
+    return signedToken;
+  }
+  const {clientId} = initOptions;
+  var userId = window.localStorage.getItem('userId');
+  if (!userId) {
+    userId = generateUid();
+    window.localStorage.setItem('userId', userId);
+  }
+  var screenSize = {width: window.screen.width, height: window.screen.height};
+  var deviceType = window.navigator.userAgent.match(/Android|BlackBerry|iPhone|iPad|iPod|Opera Mini|IEMobile/i) ? 'Mobile' : 'Desktop';
+  var e = module.init();
+  if (!ua['userId']) {
+    ua['userId'] = userId;
+    ua['client_id'] = clientId;
+    ua['plateform_name'] = e.os.name;
+    ua['os_version'] = e.os.version;
+    ua['browser_name'] = e.browser.name;
+    ua['browser_version'] = e.browser.version;
+    ua['screen_size'] = screenSize;
+    ua['device_type'] = deviceType;
+    ua['time_zone'] = window.Intl.DateTimeFormat().resolvedOptions().timeZone;
+  }
+  var signedToken = getJWToken(ua);
```

Comment:
@Prebid-bydata what is the reasoning behind using JWT? (especially that secret key is not really secret)

Reply:
hi @aleksatr , we will use this in future , currently backend requirement.